src/llmcompressor/pipelines/sequential/pipeline.py

```diff
@@ -0,0 +1,72 @@
+from typing import List
+
+import torch
+import torch.utils.data.dataloader
+import tqdm
+from compressed_tensors.utils import get_execution_device
+
+from llmcompressor.modifiers.utils.hooks import HooksMixin
+from llmcompressor.pipelines.cache import IntermediatesCache
+from llmcompressor.pipelines.sequential.helpers import trace_subgraphs
+from llmcompressor.utils.helpers import calibration_forward_context
+
+__all__ = ["run_pipeline"]
+
+
+def run_pipeline(
+    model: torch.nn.Module,
+    sequential_targets: List[str],
+    ignore: List[str],
+    dataloader: torch.utils.data.DataLoader,
+):
+    """
+    Run a sequential data pipeline according to the following steps:
+
+    1. The model is partitioned into subgraphs according to `sequential_targets`
+    2. Data passes through each subgraph sequentially. Data is passed through each
+        subgraph twice, once to trigger calibration hooks, then a second time in order
+        to capture activations after quantization has occurred through the hooks.
+    3. The intermediate activations between each subgraph are cached and offloaded to
+        the cpu between each batch in order to save memory
+
+    This pipeline requires that the model be traceable with respect to data from the
+    data loader. This may be an issue for vision language models with vision datasets,
+    due to specialized input processing in the model.
+
+    In the event that tracing fails, a torch.fx.proxy.TraceError will be raised. A model
+    can be made traceable by wrapping the untraceable functions (see
+    llmcompressor.transformers.tracing)
+    """
+    # trace subgraphs
+    sample_input = next(iter(dataloader))
+    subgraphs = trace_subgraphs(model, sample_input, sequential_targets, ignore)
+
+    with calibration_forward_context(model):
+        # prepare intermediates cache
+        model_device = get_execution_device(model)
+        intermediates = IntermediatesCache.from_dataloader(dataloader, model_device)
```

Comment:
Why do we not need to capture the activations before the first subgraph, as we have to do with raising the early exception for the layer_sequential pipeline? Seems like the IntermediateCache is filled with values from the dataloader, not any previous activations.

Reply:
The layer sequential pipeline processes everything before the first layer and captures the activations.

The sequential pipeline assigns everything before the first layer to the first subgraph. Said another way, all the processing that is done by `capture_first_layer_intermediates` is done in the sequential pipeline by processing the first subgraph.

```
subgraph[0] ~= capture_first_layer_intermediates
subgraph[1] ~= layer[0]
```